org.eclipse.jdt.ui/core refactoring/org/eclipse/jdt/internal/corext/refactoring/util/SideEffectChecker.java

```diff
@@ -0,0 +1,332 @@
+			if (MARKEDMETHODSET.contains(getQualifiedName(resolveMethodBinding))
+					|| resolveMethodBinding == null) {
+				fSideEffect= true;
+				return false;
+			}
+			MethodDeclaration md= findFunctionDefinition(resolveMethodBinding.getDeclaringClass(), resolveMethodBinding);
+			if (md != null && md.getLength() < THRESHOLD) {
+				MethodVisitor mv= new MethodVisitor();
+				md.accept(mv);
+				fSideEffect= mv.hasUpdateNoTemp() == true ? true : fSideEffect;
+			}
+		}
+		return super.preVisit2(node);
+	}
+
+	private String getQualifiedName(IMethodBinding imb) {
+		if (imb == null || imb.getDeclaringClass() == null)
+			return null;
+		return imb.getDeclaringClass().getQualifiedName() + "." + imb.getName(); //$NON-NLS-1$
+	}
+
+	private MethodDeclaration findFunctionDefinition(ITypeBinding iTypeBinding, IMethodBinding methodBinding) {
+		if (methodBinding == null || iTypeBinding == null) {
+			fSideEffect= true;
+			return null;
+		}
+		if (!(iTypeBinding.getJavaElement() instanceof IType))
+			return null;
+		IType it= (IType) (iTypeBinding.getJavaElement());
+		try {
+			IJavaElement root= it.getAncestor(IJavaElement.PACKAGE_FRAGMENT_ROOT);
+			if (root instanceof IPackageFragmentRoot) {
+				IClasspathEntry cpEntry= ((IPackageFragmentRoot) root).getRawClasspathEntry();
+				if (cpEntry.getEntryKind() == IClasspathEntry.CPE_CONTAINER
+						&& cpEntry.getPath().toString().startsWith("org.eclipse.jdt.launching.JRE_CONTAINER")) { //$NON-NLS-1$
+					return null;
+				}
+			}
+			ITypeHierarchy ith= it.newTypeHierarchy(iTypeBinding.getJavaElement().getJavaProject(), null);
+			IMethod iMethod= (IMethod) methodBinding.getJavaElement();
+			if (iMethod == null || ith == null) {
+				fSideEffect= true;
+				return null;
+			}
+
+			ArrayList<IType> iTypes= new ArrayList<>();
+			findTypes(it, ith, iTypes);
+			MethodDeclaration res= null;
+			for (IType t : iTypes) {
+				IMethod tmp= JavaModelUtil.findMethod(iMethod.getElementName(),
+						iMethod.getParameterTypes(), false, t);
+				if (tmp != null && res == null) {
+					ICompilationUnit icu= tmp.getCompilationUnit();
+					if (icu == null || icu.getSource() == null) {
+						return null;
+					}
+					ASTParser parser= ASTParser.newParser(AST.getJLSLatest());
+					parser.setKind(ASTParser.K_COMPILATION_UNIT);
+					parser.setSource(icu);
+					parser.setResolveBindings(true);
+					CompilationUnit compilationUnit= (CompilationUnit) parser.createAST(null);
+					final ASTNode perform= NodeFinder.perform(compilationUnit, tmp.getSourceRange());
+					if (perform instanceof MethodDeclaration && ((MethodDeclaration) perform).resolveBinding() != null) {
+						MethodDeclaration md= (MethodDeclaration) perform;
+						if (Modifier.isAbstract(md.resolveBinding().getModifiers()))
+							continue;
+						res= md;
+					} else {
+						fSideEffect= true;
+						return null;
+					}
+				} else if (tmp != null && res != null) {
+					return null;
```

Comment:
What scenario does this represent where you don't set sideEffect to true?

Reply:
It represents the scenario for multiple possible function implementations due to run-time polymorphism in Java. A typical example is:
```java
Person p = new Man();   
p.toString();
class Person{
    public String toString() {
        String name = "Person";
        return name;
    }
}
 
class Man extends Person{
    public String toString(){
        String name = "Man";
        return name;
    }
}
```
`p.toString()` executes `toString()` in the `Man` class rather than `Person` class. Considering this uncertainty, we adopt a conservative strategy, which is considered to have no side effects. In addition, there are some other strategies that can be used:
1. Scan for all potential methods. This strategy is the most accurate but significantly increases the computational cost.
2. Only choose one of the methods for detection. The overriding methods have similar effects that guarantee a certain accuracy.
3. Marked as having side effects. It can guarantee enough security, but it may cause a bad user experience because of the common use of polymorphism.

Reply:
I don't think that an overridden method should be assumed to be side-effect free which is what you are doing by returning null.  I think option 2 is a reasonable assumption but you should check for an abstract method and continue on until you find a real implementation.